### PR TITLE
[SIEM] Office 365 module

### LINKED ISSUE #16196: [Meta] Module to ingest Office365 audit events
* [x] o365audit input ([draft](https://github.com/elastic/beats/pull/16244))
* [ ] persist state (needs #14229)
* [x] module with filesets for different types (Azure, Exchange, Sharepoint..) (https://github.com/elastic/beats/pull/16386)
* [ ] (tbd) mock server for integration tests
* [ ] Dashboards
 
## Modules

For a metricset to go GA, the following criterias should be met:

* [x] Supported versions are documented
* [x] Supported operating systems are documented (if applicable)
* [ ] Integration tests exist
* [ ] System tests exist
* [ ] Automated checks that all fields are documented
* [ ] Documentation
* [ ] Fields follow [ECS](https://github.com/elastic/ecs) and [naming conventions](https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html)
* [ ] Dashboards exists (if applicable)
* [ ] Kibana Home Tutorial (if applicable)
  * Open PR against Kibana repo with tutorial. Examples can be found [here](https://github.com/elastic/kibana/tree/master/src/legacy/core_plugins/kibana/server/tutorials).

## Filebeat module

* [ ] Test log files exist for the grok patterns
* [ ] Generated output for at least 1 log file exists

## Metricbeat module

* [ ] Example `data.json` exists and an automated way to generate it exists (`go test -data`)
* [ ] Test environment in Docker exist for integration tests

### PR DESCRIPTION
**Pull request by adriansr**
## What does this PR do?

This includes a new fileset, o365.audit, that uses the o365audit input
to ingest logs using the Office 365 Management API.

## TODO

- [x] Map more API types.
- [x] Validate ECS mappings for SIEM app. (Looking good to me.)
- [x] Documentation.
- [x] Simplified testing logs.
- [x] Dashboard.

- Closes #16196
- Depends #16244

### CODE REVIEW

**Comment by adriansr** [comment]
Forgot to make it a draft PR.

This includes the changes on https://github.com/elastic/beats/pull/16244

File: x-pack/filebeat/module/o365/audit/config/pipeline.js [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+        to: 'o365audit.ExtendedProperties',
+    }));
+    builder.Add("makeModifiedPropertyDict", makeDictFromModifiedPropertyArray({
+        from: 'o365audit.ModifiedProperties',
+        to: 'o365audit.ModifiedProperties',
+    }));
+
+    // Populate event specific fields.
+    builder.Add("productSpecific", makeConditional({
+     condition: function(event) {
```
Comment (andrewkroh) [inline x-pack/filebeat/module/o365/audit/config/pipeline.js]: Indentation is different than the rest.

File: x-pack/filebeat/module/o365/audit/config/pipeline.js [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+        target: 'event',
+        fields: {
+            kind: 'event',
+        },
+    }));
+    builder.Add("setUserFieldsFromId", new processor.Dissect({
+        tokenizer: "%{name}@%{domain}",
+        field: "user.id",
+        target_prefix: "user",
+        'when.contains.user.id': '@',
```
Comment (andrewkroh) [inline x-pack/filebeat/module/o365/audit/config/pipeline.js]: Should we also try to parse the other form, like `SHAREPOINT\system`, where we have the user.domain and user.name? I also see some have a form like `NT AUTHORITY\\SYSTEM (Microsoft.Exchange.ServiceHost)`, but I'm not sure what the `Microsoft.Exchange.ServiceHost` means.

**Comment by andrewkroh** [comment]
@leehinman can you please take a look at the ECS mappings in this PR.

**Comment by adriansr** [comment]
@andrewkroh @leehinman this is still a WIP (should be a draft, I clicked the wrong button and made it a PR), but I will push the final mappings soon.

**Comment by adriansr** [comment]
Ready for final reviews / ECS validation. I've mapped all the API schemas that made sense to me. Please have a look only at the last commit (`483e3bb22`) as the rest is just the o365audit input which is in review at #16244.

Docs / dashboards will be added on a separate PR after this is merged.

File: x-pack/filebeat/module/o365/audit/config/pipeline.js (hunk 2 above)
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/config/pipeline.js]: These are all the different "user IDs" that I've captured:

```
ASR@TESTSIEM2.ONMICROSOFT.COM
DlpAgent
NT AUTHORITY\SYSTEM (Microsoft.Exchange.ServiceHost)
S-1-5-18
SecurityComplianceAlerts
Service Account
ServicePrincipal_148657c9-73a1-4f49-8a6d-051984dc9d5e
ServicePrincipal_e2aa77b1-3ac1-4bce-9f7e-cd2440b8e5c2
Unknown
alice@testsiem.onmicrosoft.com
alice@testsiem2.onmicrosoft.com
app@sharepoint
asr@testsiem.onmicrosoft.com
bob@testsiem.onmicrosoft.com
fim_password_service@support.onmicrosoft.com
Application
36787265537
36085768193
```
Different services seem to have a wildly different concept of an user ID. "NT AUTHORITY" is used in all messages from the "ExchangeAdmin" schema, and only in those.  I think it's better to not set "user.id" for those messages. WDYT?

File: x-pack/filebeat/module/o365/audit/test/22-yammer.log-expected.json [hunk 3]
```diff
@@ -0,0 +1,110 @@
+[
+    {
+        "@timestamp": "2020-02-28T09:42:45.000Z",
+        "client.address": "79.159.10.151",
+        "client.ip": "79.159.10.151",
+        "event.action": "GroupCreation",
```
Comment (leehinman) [inline x-pack/filebeat/module/o365/audit/test/22-yammer.log-expected.json]: groupcreate would probably be event.category = iam, event.type = ["create", "group"]

File: x-pack/filebeat/module/zeek/connection/test/connection-json.log-expected.json [hunk 4]
```diff
@@ -165,4 +165,4 @@
         "zeek.connection.state_message": "No SYN seen, just midstream traffic (a 'partial connection' that was not later closed).",
         "zeek.session_id": "Cc6NJ3GRlfjE44I3h"
     }
-]
+]
```
Comment (leehinman) [inline x-pack/filebeat/module/zeek/connection/test/connection-json.log-expected.json]: Probably shouldn't be part of this PR.

File: x-pack/filebeat/module/o365/audit/_meta/fields.yml [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+    - name: Comments
+      type: text
+      norms: false
+
+    - name: CorrelationId
+      type: keyword
+
+    - name: CreationTime
+      type: keyword
```
[the remark this answers is not in the document]
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/_meta/fields.yml]: This is the original date from the API, which is already parsed to date as `@timestamp`. I'd like to keep the fields under `o365.audit` as raw as possible.

File: x-pack/filebeat/module/o365/audit/_meta/fields.yml [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+
+    - name: ClientInfoString
+      type: keyword
+
+    - name: ClientIP
+      type: keyword
+
+    - name: ClientIPAddress
+      type: keyword
```
[the remark this answers is not in the document]
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/_meta/fields.yml]: In that case we risk that if the API provides something that is not a valid IP, it won't be possible to ingest the event. This is already converted to type ip in `client.ip`.

File: x-pack/filebeat/module/o365/audit/_meta/fields.yml [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    - name: Category
+      type: keyword
+
+    - name: ClientAppId
+      type: keyword
+
+    - name: ClientInfoString
+      type: keyword
+
+    - name: ClientIP
```
[the remark this answers is not in the document]
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/_meta/fields.yml]: Not 100% sure.

ClientIP: 

> The IP address of the device that was used when the activity was logged. The IP address is displayed in either an IPv4 or IPv6 address format.
>
> For some services, the value displayed in this property might be the IP address for a trusted application (for example, Office on the web apps) calling into the service on behalf of a user and not the IP address of the device used by person who performed the activity.
>
> Also, for Azure Active Directory-related events, the IP address isn't logged and the value for the ClientIP property is null.

ClientIPAddress:

> The IP address of the device that was used when the operation was logged. The IP address is displayed in either an IPv4 or IPv6 address format.

What I did is set `client.ip` from both. Keeping the value from `ClientIPAddress` if it exists, as I understand sometimes `ClientIP` can be the IP of an intermediate service.

File: x-pack/filebeat/module/o365/audit/config/pipeline.js [hunk 8]
```diff
@@ -0,0 +1,786 @@
...
+        },
+    }));
+
+    builder.Add("saveFields", new processor.Convert({
+        fields: [
+            // SharePoint metadata
+            {from: 'o365audit.SharePointMetaData.From', to: 'user.id'},
+            {from: 'o365audit.SharePointMetaData.FileName', to: 'file.name'},
+            {from: 'o365audit.SharePointMetaData.FilePathUrl', to: 'url.original'},
+            {from: 'o365audit.SharePointMetaData.UniqueId', to: 'file.inode'},
```
[the remark this answers is not in the document]
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/config/pipeline.js]: It's documented as `UniqueId` but the API is producing `UniqueID`. Was just being a little too defensive in here, just in case they realize the discrepancy and change the generated field instead of the docs.

File: x-pack/filebeat/module/o365/audit/config/pipeline.js [hunk 9]
```diff
@@ -0,0 +1,786 @@
...
+            evt.Put("rule.id", ruleIds[0]);
+            evt.Put("rule.name", ruleNames[0]);
+        } else if (ruleIds.length > 0) {
+            evt.Put("rule.id", ruleIds);
+            evt.Put("rule.name", ruleNames);
+        }
+        if (maxSeverity > -1) {
+            evt.Put("event.severity", maxSeverity);
+        }
+        evt.Put("event.outcome", allowed ||  isBlockOverride(evt)? "success" : "failure");
```
[the remark this answers is not in the document]
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/config/pipeline.js]: The operator precedence of javascript is different from what you might expect. This expression is equivalent to:
> ( allowed || isBlockOverride(evt) )? "success" : "failure"

I'll wrap it parentheses around it to make it more clear.

File: x-pack/filebeat/module/o365/audit/test/01-exchange-admin.log-expected.json [hunk 10]
```diff
@@ -0,0 +1,5110 @@
...
+        "o365.audit.UserId": "NT AUTHORITY\\SYSTEM (Microsoft.Exchange.ServiceHost)",
+        "o365.audit.UserKey": "NT AUTHORITY\\SYSTEM (Microsoft.Exchange.ServiceHost)",
+        "o365.audit.UserType": 3,
+        "o365.audit.Version": 1,
+        "o365.audit.Workload": "Exchange",
+        "organization.id": "b86ab9d4-fcf1-4b11-8a06-7a8f91b47fbd",
+        "organization.name": "testsiem.onmicrosoft.com",
+        "server.address": "HE1PR0102MB3228 (15.20.2707.017)",
+        "service.type": "o365",
+        "user.hash": "NT AUTHORITY\\SYSTEM (Microsoft.Exchange.ServiceHost)",
```
[the remark this answers is not in the document]
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/test/01-exchange-admin.log-expected.json]: Right. I was too eager to try to map as many fields as possible. The source field "UserKey" is really just an alternate user ID (also with different meanings across systems). I think it's better not to map it to ECS at all.

File: x-pack/filebeat/module/o365/audit/test/11-dlp-sharepoint.log-expected.json [hunk 11]
```diff
@@ -0,0 +1,605 @@
...
+        "event.id": "a21f13b9-22b6-405b-bf9e-a07ad8d456da",
+        "event.kind": "alert",
+        "event.module": "o365",
+        "event.outcome": "success",
+        "event.provider": "OneDrive",
+        "event.severity": 2,
+        "event.type": "access",
+        "file.inode": "9cc7be1c-dd5a-4895-b7cb-757de6d51b42",
+        "file.name": "Customers Financial Data.docx",
+        "file.owner": "Alan Smithee",
```
[the remark this answers is not in the document]
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/test/11-dlp-sharepoint.log-expected.json]: Good idea. Done.

File: x-pack/filebeat/module/o365/audit/test/22-yammer.log-expected.json (hunk 3 above)
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/test/22-yammer.log-expected.json]: Thanks for the suggestion. I wasn't aware of this new category value.

File: x-pack/filebeat/module/zeek/connection/test/connection-json.log-expected.json (hunk 4 above)
Comment (adriansr) [reply x-pack/filebeat/module/zeek/connection/test/connection-json.log-expected.json]: Good catch

File: x-pack/filebeat/module/o365/audit/config/pipeline.js [hunk 12]
```diff
@@ -0,0 +1,852 @@
...
+}
+
+function azureADLogonSchema(debug) {
+    var builder = new PipelineBuilder("o365.audit.AzureActiveDirectory", debug);
+    builder.Add("setEventAuthFields", function(evt){
+       evt.Put("event.category", "authentication");
+       var outcome = evt.Get("event.outcome");
+       // As event.type is an array, this sets both the traditional
+       // "authentication_success"/"authentication_failure"
+       // and the soon to be ECS standard "start".
```
Comment (andrewkroh) [inline x-pack/filebeat/module/o365/audit/config/pipeline.js]: `event.type: start` is no longer "soon to be" 😄 .

**Comment by adriansr** [comment]
jenkins, test this

File: x-pack/filebeat/module/o365/audit/test/15-azuread-sts-logon.log-expected.json [hunk 13]
```diff
@@ -0,0 +1,6350 @@
...
+        "event.code": "AzureActiveDirectoryStsLogon",
+        "event.dataset": "o365.audit",
+        "event.id": "ca0efc24-1b89-4962-8fef-a3ac5437302f",
+        "event.kind": "event",
+        "event.module": "o365",
+        "event.outcome": "success",
+        "event.provider": "AzureActiveDirectory",
+        "event.type": [
+            "start",
+            "authentication_success"
```
Comment (leehinman) [inline x-pack/filebeat/module/o365/audit/test/15-azuread-sts-logon.log-expected.json]: Question.  I don't think "authentication_success" is an official ECS event.type. I know it has been used in the past, but do we want to keep using it?

File: x-pack/filebeat/module/o365/audit/test/22-yammer.log-expected.json [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        "event.category": "iam",
+        "event.code": "Yammer",
+        "event.dataset": "o365.audit",
+        "event.id": "2af7bbf1-d5d8-5cb0-8aca-f4ad8a087594",
+        "event.kind": "event",
+        "event.module": "o365",
+        "event.outcome": "success",
+        "event.provider": "Yammer",
+        "event.type": "group",
```
Comment (leehinman) [inline x-pack/filebeat/module/o365/audit/test/22-yammer.log-expected.json]: would it be possible to have "creation" added to event.type.  So it would be event.cateogory = iam, event.type = "creation, group"

File: x-pack/filebeat/module/o365/audit/test/15-azuread-sts-logon.log-expected.json (hunk 13 above)
Comment (adriansr) [reply x-pack/filebeat/module/o365/audit/test/15-azuread-sts-logon.log-expected.json]: I just added it because I thought the SIEM app would still be looking for it. In the future we can remove it from here and other places where it's used (auditbeat's system/login).

Comment (leehinman) [reply x-pack/filebeat/module/o365/audit/test/15-azuread-sts-logon.log-expected.json]: makes sense.  thanks.
